python/cudf/cudf/core/spill_manager.py

```diff
@@ -0,0 +1,368 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import gc
+import io
+import os
+import threading
+import traceback
+import warnings
+import weakref
+from dataclasses import dataclass
+from functools import cached_property
+from typing import List, Mapping, MutableMapping, Optional, Set, Tuple
+
+import rmm.mr
+
+from cudf._lib.column import Column
+from cudf.core.buffer import DeviceBufferLike, as_device_buffer_like
+from cudf.core.column_accessor import ColumnAccessor
+from cudf.core.frame import Frame
+from cudf.core.indexed_frame import IndexedFrame
+from cudf.core.spillable_buffer import SpillableBuffer
+from cudf.utils.string import format_bytes
+
+
+def get_traceback() -> str:
+    with io.StringIO() as f:
+        traceback.print_stack(file=f)
+        f.seek(0)
+        return f.read()
+
+
+@dataclass
+class ExposeStatistic:
+    traceback: str
+    count: int = 1
+    total_nbytes: int = 0
+    spilled_nbytes: int = 0
+
+
+class SpillManager:
+    _base_buffers: MutableMapping[int, SpillableBuffer]
+    _other_buffers: MutableMapping[int, DeviceBufferLike]
+    _expose_statistics: Optional[MutableMapping[str, ExposeStatistic]]
+
+    def __init__(
+        self,
+        *,
+        spill_on_demand=False,
+        device_memory_limit=None,
+        expose_statistics=False,
+    ) -> None:
+        self._lock = threading.Lock()
+        self._base_buffers = weakref.WeakValueDictionary()
+        self._other_buffers = weakref.WeakValueDictionary()
+        self._id_counter = 0
+        self._spill_on_demand = spill_on_demand
```

Comment:
>  if spill_on_demand=False, we only spill whenever a SpillableBuffer is created that overflows the set memory limit, so that only really supports the case where all GPU memory allocated in a process is in the form of a SpillableBuffer.

Yes, that is correct.
Normally, we want `spill_on_demand=True`, which is also what `GlobalSpillManager` does by default (can be overwritten by `CUDF_SPILL_ON_DEMAND`). However, I don't think that just the creation of `SpillManager` should register the out-of-memory handler.

Reply:
A compromise: `oom()` become the method `._out_of_memory_handle()` and `.register_spill_on_demand()` is removed.